src/bufr/BufrParser/Query/Tokenizer.h

```diff
@@ -0,0 +1,386 @@
+        {
+            static const auto regex = std::regex(T::Pattern);
+            std::shared_ptr<T> token = nullptr;
+            std::smatch matches;
+            if (std::regex_search(start, end, matches, regex))
+            {
+                if (matches.position() == 0)
+                {
+                    start += matches[0].str().size();
+                    token = std::make_shared<T>();
+                    token->str_ = matches[0].str();
+                    token->tokenize();
+                }
+            }
+
+            return token;
+        }
+
+        std::string debugStr() const override { return T::DebugStr; }
+    };
+
+    /// \brief Token for the seperator between query components (i.e. '/')
+    class SeparatorToken : public TokenBase<SeparatorToken> {
+     public:
+        constexpr static const char* Pattern = "\\/";
+        constexpr static const char* DebugStr = "<sep>";
+    };
+
+    /// \brief Token for the wildcard subset
+    class AnySubset : public TokenBase<AnySubset> {
+     public:
+        constexpr static const char* Pattern = "\\*";
+        constexpr static const char* DebugStr = "<anySubset>";
+    };
+
+    /// \brief Token for the index part of a query (i.e. '[2]')
+    class IndexToken : public TokenBase<IndexToken> {
+     public:
+        constexpr static const char* Pattern = "\\[[0-9]+\\]";
+        constexpr static const char* DebugStr = "<index>";
+
+        // \brief Get the debug string for the token.
+        std::string debugStr() const override
+        {
+            std::ostringstream debugStr;
+            debugStr << "<index=" << index() << ">";
+            return debugStr.str();
+        }
+
+        size_t index() const
+        {
+            return std::stoi(std::string(++str_.begin(), --str_.end()));
+        }
+    };
+
+    /// \brief Token for the mnemonic parts of a query (i.e. 'TMBR')
+    class MnemonicToken : public TokenBase<MnemonicToken> {
+     public:
+        constexpr static const char* Pattern = "([A-Z0-9_]+)";
+        constexpr static const char* DebugStr = "<mnemonic>";
+
+        // \brief Get the debug string for the token.
+        std::string debugStr() const override
+        {
+            std::ostringstream debugStr;
+            debugStr << "<mnemonic=" << str_ << ">";
+            return debugStr.str();
+        }
+    };
+
+    /// \brief Token for the filter parts of a query (i.e. '{1-3,5}')
+    class FilterToken : public TokenBase<FilterToken> {
+     public:
+        constexpr static const char* Pattern = "\\{\\d+(\\-\\d+)?(\\,\\d+(\\-\\d+)?)*\\}";
+        constexpr static const char* DebugStr = "<filter>";
+
+        std::vector<size_t> indices() { return indices_; }
+
+        /// \brief Tokenize the str_ into tokens.
+        void tokenize() final
+        {
+            static const auto filterRegex =
+                std::regex("\\{(\\d+(\\-\\d+)?)(,\\d+(\\-\\d+)?)*\\}");
```

Comment:
this line does not have  `\\` in the second set of parenthesis different from the Pattern constexpr line above
```suggestion
                std::regex("\\{(\\d+(\\-\\d+)?)(\\,\\d+(\\-\\d+)?)*\\}");
```

Reply:
It's actually more correct to remove the `\\` in front of the comma as commas don't need to be escaped (are not treated as special characters in regex).